packages/indexer/src/indexers/amm-market/amm-market-instruction-indexer.ts

```diff
@@ -0,0 +1,171 @@
+  SwapType,
+} from "@metadaoproject/futarchy-ts";
+import { InstructionIndexer } from "../instruction-indexer";
+import { IDL } from "@openbook-dex/openbook-v2";
+import { SolanaParser } from "@debridge-finance/solana-transaction-parser";
+
+export enum AmmInstructionIndexerError {
+  GeneralError = "GeneralError",
+  MissingMarket = "MissingMarket",
+}
+
+const ammClient = new AmmClient(provider, AMM_PROGRAM_ID, []);
+
+type IDL = typeof ammClient.program.idl;
+
+// create handle new instruction function, which makes a call to store the transaction in the DB and
+// then gets the instruction indexer implementation for the account...
+
+const ammParser = new SolanaParser([
+  {
+    idl: ammClient.program.idl,
+    programId: AMM_PROGRAM_ID,
+  },
+]);
+
+export const AmmMarketInstructionsIndexer: InstructionIndexer<IDL> = {
+  PROGRAM_ID: AMM_PROGRAM_ID.toString(),
+  PROGRAM_IDL: ammClient.program.idl,
+  PROGRAM_NAME: "amm",
+  indexInstruction: async (
+    transactionIndex: number,
+    transactionResponse: VersionedTransactionResponse,
+    instructionIndex: number,
+    decodedInstruction: IDL["instructions"][number]
+  ) => {
+    try {
+      return Ok({ txSig: "" });
+    } catch (e) {
+      console.error(e);
+      return Err({ type: AmmInstructionIndexerError.GeneralError });
+    }
+  },
+  async indexTransactionSig(transaction: TransactionRecord): Promise<
+    Result<
+      {
+        txSig: string;
+      },
+      TaggedUnion
+    >
+  > {
+    try {
+      const ixs = await ammParser.parseTransaction(
+        connection,
+        transaction.txSig
+      );
+      const ixTypes = ixs?.reduce(
+        (prev, currIx) => {
+          // Check for 'swap' instruction
+          if (currIx.name === "swap") {
+            prev.swap = true;
+          }
+          // Check for 'addLiquidity' instruction
+          else if (currIx.name === "addLiquidity") {
+            prev.addLiquidity = true;
+          }
+          // Check for 'removeLiquidity' instruction
+          else if (currIx.name === "removeLiquidity") {
+            prev.removeLiquidity = true;
+          }
+
+          return prev;
+        },
+        { swap: false, addLiquidity: false, removeLiquidity: false }
+      );
+      if (ixTypes?.swap) {
+        const relatedIx = ixs?.find((i) => i.name === "swap");
+        const marketAcct = relatedIx?.accounts.find((a) => a.name === "amm");
+        if (!marketAcct) return Err({ type: "missing data" });
+        const userAcct = relatedIx?.accounts.find((a) => a.name === "user");
+        if (!userAcct) return Err({ type: "missing data" });
+        if (!relatedIx?.args) return Err({ type: "missing data" });
+        const args = relatedIx.args as {
+          args: {
+            swapType: SwapType;
+            inputAmount: BN;
+            outputAmountMin: BN;
+          };
+        };
+        // index a swap here
+        const swapOrder: OrdersRecord = {
+          marketAcct: marketAcct.pubkey.toBase58(),
```

Comment:
String I think

Reply:
check my above comment on this. This method may be "better"

Reply:
interesting one: https://solana.stackexchange.com/questions/3514/publickey-tostring-vs-publickey-tobase58

Reply:
lulz